sixtyfps_runtime/corelib/model.rs

```diff
@@ -64,18 +64,92 @@ impl ModelNotify {
+///     // the backing data, stored in a `RefCell` as this model can be modified
+///     array: std::cell::RefCell<Vec<T>>,
+///     // the ModelNotify will allow to notify the UI that the model changes
+///     notify: ModelNotify,
+/// }
+///
+/// impl<T: Clone + 'static> Model for VecModel<T> {
+///     type Data = T;
+///
+///     fn row_count(&self) -> usize {
+///         self.array.borrow().len()
+///     }
+///
+///     fn row_data(&self, row: usize) -> Self::Data {
+///         self.array.borrow()[row].clone()
+///     }
+///
+///     fn set_row_data(&self, row: usize, data: Self::Data) {
+///         self.array.borrow_mut()[row] = data;
+///         // don't forget to call row_changed
+///         self.notify.row_changed(row);
+///     }
+///
+///     fn attach_peer(&self, peer: ModelPeer) {
+///         // simply forward to ModelNotify::attach
+///         self.notify.attach(peer);
+///     }
+///
+///     fn as_any(&self) -> &dyn core::any::Any {
+///         // a typical implementation just return `self`
+///         self
+///     }
+/// }
+///
+/// // when modifying the model, we call the corresponding function in
+/// // the ModelNotify
+/// impl<T> VecModel<T> {
+///     /// Add a row at the end of the model
+///     pub fn push(&self, value: T) {
+///         self.array.borrow_mut().push(value);
+///         self.notify.row_added(self.array.borrow().len() - 1, 1)
+///     }
+///
+///     /// Remove the row at the given index from the model
+///     pub fn remove(&self, index: usize) {
+///         self.array.borrow_mut().remove(index);
+///         self.notify.row_removed(index, 1)
+///     }
+/// }
+/// ```
 pub trait Model {
     /// The model data: A model is a set of row and each row has this data
     type Data;
     /// The amount of row in the model
     fn row_count(&self) -> usize;
     /// Returns the data for a particular row. This function should be called with `row < row_count()`.
     fn row_data(&self, row: usize) -> Self::Data;
-    /// Sets the data for a particular row. This function should be called with `row < row_count()`.
-    /// If the model cannot support data changes, then it is ok to do nothing (default implementation).
-    /// If the model can update the data, it should also call row_changed on its internal `ModelNotify`.
-    fn set_row_data(&self, _row: usize, _data: Self::Data) {}
-    /// Should forward to the internal [`ModelNotify::attach`]
+    /// Sets the data for a particular row.
+    ///
+    /// This function should be called with `row < row_count()`, otherwise the implementation can panic.
+    ///
+    /// If the model cannot support data changes, then it is ok to do nothing.
+    /// The default implementation will print a warning to stderr.
+    ///
+    /// If the model can update the data, it should also call [`ModelNofity::row_changed`] on its
+    /// internal [`ModelNotify`].
+    fn set_row_data(&self, _row: usize, _data: Self::Data) {
+        eprintln!(
```

Comment:
Maybe this could use debug_log! so that in wasm build it prints on console.log? Just a thought, doesn't have to been this change either.